# Challenge Alura - TelecomX

## 📌 Extracción

Para iniciar tu análisis, necesitarás importar los datos de la API de Telecom X. Estos datos están disponibles en formato JSON y contienen información esencial sobre los clientes, incluyendo datos demográficos, tipo de servicio contratado y estado de evasión.

#### ✅ Cargar los datos directamente desde la API utilizando Python.

In [164]:
import pandas as pd

# Cargar los datos directamente desde la API utilizando Python
url = 'https://raw.githubusercontent.com/alura-cursos/challenge2-data-science-LATAM/main/TelecomX_Data.json'

# url = 'TelecomX_Data.json' 

#### ✅ Convertir los datos a un DataFrame de Pandas para facilitar su manipulación.

In [165]:
try:
    # Convertir los datos a un DataFrame de Pandas
    Datos = pd.read_json(url)
    print(f"DataFrame cargado exitosamente. Filas: {len(Datos)}, Columnas: {len(Datos.columns)}")
except Exception as e:
    print(f"Error al cargar el JSON: {e}")
    print("Verifica la URL y tu conexión a internet.")

DataFrame cargado exitosamente. Filas: 7267, Columnas: 6


In [166]:
Datos.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

#### ✅ Normalización de las columnas | Archivo JSON

In [167]:
# Datos anidados 
# Columnas [customer, phone, internet, account]
Datos_Norm_customer = pd.json_normalize(Datos['customer'])
Datos_Norm_phone = pd.json_normalize(Datos['phone'])
Datos_Norm_internet = pd.json_normalize(Datos['internet'])
Datos_Norm_account = pd.json_normalize(Datos['account'])

In [168]:
#Forma de concatenar Series
#Nuevo DF
Datos_Norm = pd.concat([Datos['customerID'], Datos['Churn'], Datos_Norm_customer, Datos_Norm_phone,Datos_Norm_internet, Datos_Norm_account], axis=1)
Datos_Norm.head(5)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9          267.4  

[5 rows x 21 columns]

## 🔧 Transformación

### Conoce el conjunto de datos

Ahora que has extraído los datos, es fundamental comprender la estructura del dataset y el significado de sus columnas. Esta etapa te ayudará a identificar qué variables son más relevantes para el análisis de evasión de clientes.

#### ¿Qué debes hacer?

##### ✅ Explorar las columnas del dataset y verificar sus tipos de datos.

In [216]:
# Anterior df
Datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


In [217]:
# Anterior df
Datos.dtypes

customerID    object
Churn         object
customer      object
phone         object
internet      object
account       object
dtype: object

In [218]:
#Nuevo df
Datos_Norm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   customerid        7267 non-null   string  
 1   churn             7267 non-null   category
 2   gender            7267 non-null   category
 3   seniorcitizen     7267 non-null   bool    
 4   partner           7267 non-null   category
 5   dependents        7267 non-null   category
 6   tenure            7267 non-null   int64   
 7   phoneservice      7267 non-null   category
 8   multiplelines     7267 non-null   category
 9   internetservice   7267 non-null   category
 10  onlinesecurity    7267 non-null   category
 11  onlinebackup      7267 non-null   category
 12  deviceprotection  7267 non-null   category
 13  techsupport       7267 non-null   category
 14  streamingtv       7267 non-null   category
 15  streamingmovies   7267 non-null   category
 16  contract          7267 n

In [219]:
Datos_Norm.dtypes

customerid          string[python]
churn                     category
gender                    category
seniorcitizen                 bool
partner                   category
dependents                category
tenure                       int64
phoneservice              category
multiplelines             category
internetservice           category
onlinesecurity            category
onlinebackup              category
deviceprotection          category
techsupport               category
streamingtv               category
streamingmovies           category
contract                  category
paperlessbilling          category
paymentmethod             category
charges.monthly            float64
charges.total              float64
dtype: object

In [220]:
print(f"Datos_Norm contiene. Filas: {len(Datos_Norm)}, Columnas: {len(Datos_Norm.columns)}")


Datos_Norm contiene. Filas: 7267, Columnas: 21


##### ✅ Consultar el diccionario para comprender mejor el significado de las variables.
##### ✅ Identificar las columnas más relevantes para el análisis de evasión.

### Comprobación de incoherencias en los datos

En este paso, verifica si hay problemas en los datos que puedan afectar el análisis. Presta atención a valores ausentes, duplicados, errores de formato e inconsistencias en las categorías. Este proceso es esencial para asegurarte de que los datos estén listos para las siguientes etapas.

#### ¿Qué debes hacer?
---
##### Buscar:

#### Consistencia y validez de datos categorias

##### Duplicados

In [221]:
#Verificando si hay duplicados
duplicados = Datos_Norm.duplicated().sum() 
print(f"Numero de lineas duplicadas: {duplicados}")

Numero de lineas duplicadas: 0


##### Valores Vacios


In [222]:
for col in Datos_Norm.columns:
        vacios = (Datos_Norm[col] == "").sum()  # Cuenta los valores exactamente iguales a ""
        print(f" • {col}: {vacios} valores vacíos ('')")
        #Si el numero de valores vacios es mayor a 0, muestra ejemplos de los datos vacios
        if vacios > 0:
            print(f"Las Filas tienen datos vacios son: ", Datos_Norm[Datos_Norm[col] == ""][col].sample(5))  # Muestra ejemplos si hay vacíos
            print('-' * 50)

 • customerid: 0 valores vacíos ('')
 • churn: 224 valores vacíos ('')
Las Filas tienen datos vacios son:  4753    
3305    
5561    
3220    
2879    
Name: churn, dtype: category
Categories (3, object): ['', 'No', 'Yes']
--------------------------------------------------
 • gender: 0 valores vacíos ('')
 • seniorcitizen: 0 valores vacíos ('')
 • partner: 0 valores vacíos ('')
 • dependents: 0 valores vacíos ('')
 • tenure: 0 valores vacíos ('')
 • phoneservice: 0 valores vacíos ('')
 • multiplelines: 0 valores vacíos ('')
 • internetservice: 0 valores vacíos ('')
 • onlinesecurity: 0 valores vacíos ('')
 • onlinebackup: 0 valores vacíos ('')
 • deviceprotection: 0 valores vacíos ('')
 • techsupport: 0 valores vacíos ('')
 • streamingtv: 0 valores vacíos ('')
 • streamingmovies: 0 valores vacíos ('')
 • contract: 0 valores vacíos ('')
 • paperlessbilling: 0 valores vacíos ('')
 • paymentmethod: 0 valores vacíos ('')
 • charges.monthly: 0 valores vacíos ('')
 • charges.total: 0 valores

In [223]:
# Verificando si hay valores nulos
Datos_Norm.iloc[30]

customerid                         0047-ZHDTW
churn                                        
gender                                 Female
seniorcitizen                           False
partner                                    No
dependents                                 No
tenure                                     11
phoneservice                              Yes
multiplelines                             Yes
internetservice                   Fiber optic
onlinesecurity                            Yes
onlinebackup                               No
deviceprotection                           No
techsupport                                No
streamingtv                                No
streamingmovies                            No
contract                       Month-to-month
paperlessbilling                          Yes
paymentmethod       Bank transfer (automatic)
charges.monthly                          79.0
charges.total                           929.3
Name: 30, dtype: object

In [224]:
Datos_Norm.iloc[4541]

customerid                6225-VNQSJ
churn                               
gender                          Male
seniorcitizen                  False
partner                           No
dependents                        No
tenure                             1
phoneservice                      No
multiplelines       No phone service
internetservice                  DSL
onlinesecurity                    No
onlinebackup                      No
deviceprotection                  No
techsupport                       No
streamingtv                       No
streamingmovies                  Yes
contract              Month-to-month
paperlessbilling                 Yes
paymentmethod       Electronic check
charges.monthly                35.45
charges.total                  35.45
Name: 4541, dtype: object

Otra forma de consultar valores vacios

In [225]:
Churn_a = Datos_Norm[Datos_Norm['churn'] == '']
Churn_a.sample(3)

customerid churn gender  seniorcitizen partner dependents  tenure  \
6013  8212-FSQIV         Male          False      No        Yes       1   
1160  1645-EFCGT         Male          False     Yes        Yes      72   
5684  7766-CVTAF         Male          False     Yes        Yes      43   

     phoneservice multiplelines internetservice  ...         onlinebackup  \
6013          Yes            No              No  ...  No internet service   
1160          Yes            No              No  ...  No internet service   
5684          Yes           Yes              No  ...  No internet service   

         deviceprotection          techsupport          streamingtv  \
6013  No internet service  No internet service  No internet service   
1160  No internet service  No internet service  No internet service   
5684  No internet service  No internet service  No internet service   

          streamingmovies        contract paperlessbilling  \
6013  No internet service  Month-to-month              Yes   
1160  No internet service        Two year               No   
5684  No internet service        Two year               No   

                paymentmethod charges.monthly  charges.total  
6013             Mailed check           20.25          20.25  
1160  Credit card (automatic)           19.40        1363.25  
5684             Mailed check           24.25        1077.95  

[3 rows x 21 columns]

##### Valores Nulos

In [226]:
# Verificando se hay valores Nulos
valores_nulos = Datos_Norm.isnull().values.sum() 
print("\nValores nulos por columna: ") 
print(valores_nulos [valores_nulos > 0])


Valores nulos por columna: 
[]


In [227]:
# Formas
Datos_Norm.isna().sum()

customerid          0
churn               0
gender              0
seniorcitizen       0
partner             0
dependents          0
tenure              0
phoneservice        0
multiplelines       0
internetservice     0
onlinesecurity      0
onlinebackup        0
deviceprotection    0
techsupport         0
streamingtv         0
streamingmovies     0
contract            0
paperlessbilling    0
paymentmethod       0
charges.monthly     0
charges.total       0
dtype: int64

##### Valores Atipicos

In [228]:
for col in Datos_Norm.columns:
    únicos = Datos_Norm[col].nunique(dropna=False)
    print(f"Columna '{col}': {únicos} valores únicos")
    if únicos <= 50:    # Imprima los valores si no hay muchos
        print(Datos_Norm[col].unique())
        print('-' * 40) # Linea divisora para una mejor visualizacion

Columna 'customerid': 7267 valores únicos
Columna 'churn': 3 valores únicos
['No', 'Yes', '']
Categories (3, object): ['', 'No', 'Yes']
----------------------------------------
Columna 'gender': 2 valores únicos
['Female', 'Male']
Categories (2, object): ['Female', 'Male']
----------------------------------------
Columna 'seniorcitizen': 2 valores únicos
[False  True]
----------------------------------------
Columna 'partner': 2 valores únicos
['Yes', 'No']
Categories (2, object): ['No', 'Yes']
----------------------------------------
Columna 'dependents': 2 valores únicos
['Yes', 'No']
Categories (2, object): ['No', 'Yes']
----------------------------------------
Columna 'tenure': 73 valores únicos
Columna 'phoneservice': 2 valores únicos
['Yes', 'No']
Categories (2, object): ['No', 'Yes']
----------------------------------------
Columna 'multiplelines': 3 valores únicos
['No', 'Yes', 'No phone service']
Categories (3, object): ['No', 'No phone service', 'Yes']
-----------------------

##### Tipos de datos

In [229]:
print("▶ Tipos de datos:\n")
print(Datos_Norm.dtypes)
print("\n▶ Columnas tipo object que podrían ser numéricas:\n")
for col in Datos_Norm.select_dtypes(include='object'):
    muestra = Datos_Norm[col].dropna().astype(str).str.replace('.', '', 1).str.isnumeric()
    if muestra.sum() > 0 and muestra.sum() < len(Datos_Norm[col]):
        print(f"• {col}: contiene valores que parecen números pero están como texto.")

▶ Tipos de datos:

customerid          string[python]
churn                     category
gender                    category
seniorcitizen                 bool
partner                   category
dependents                category
tenure                       int64
phoneservice              category
multiplelines             category
internetservice           category
onlinesecurity            category
onlinebackup              category
deviceprotection          category
techsupport               category
streamingtv               category
streamingmovies           category
contract                  category
paperlessbilling          category
paymentmethod             category
charges.monthly            float64
charges.total              float64
dtype: object

▶ Columnas tipo object que podrían ser numéricas:



##### Normalizacion de texto

### Manejo de inconsistencias
Ahora que has identificado las inconsistencias, es momento de aplicar las correcciones necesarias. Ajusta los datos para asegurarte de que estén completos y coherentes, preparándolos para las siguientes etapas del análisis.


#### Manejo de tipo de datos

##### Uso de memoria Anterior

In [230]:
Datos_Norm.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   customerid        7267 non-null   string  
 1   churn             7267 non-null   category
 2   gender            7267 non-null   category
 3   seniorcitizen     7267 non-null   bool    
 4   partner           7267 non-null   category
 5   dependents        7267 non-null   category
 6   tenure            7267 non-null   int64   
 7   phoneservice      7267 non-null   category
 8   multiplelines     7267 non-null   category
 9   internetservice   7267 non-null   category
 10  onlinesecurity    7267 non-null   category
 11  onlinebackup      7267 non-null   category
 12  deviceprotection  7267 non-null   category
 13  techsupport       7267 non-null   category
 14  streamingtv       7267 non-null   category
 15  streamingmovies   7267 non-null   category
 16  contract          7267 n

##### Cambio de tipo de dato [*float*], [*string*] y [*bool*]

In [231]:
Datos_Norm['Charges.Total'] = pd.to_numeric(Datos_Norm['Charges.Total'], errors='coerce').fillna(0)
Datos_Norm['customerID'] = Datos_Norm['customerID'].astype('string')
Datos_Norm['SeniorCitizen'] = Datos_Norm['SeniorCitizen'].astype('bool')


KeyError: 'Charges.Total'

##### Cambio de tipo de dato [*Categorico*]

In [ ]:
# Columnas categóricas
columnas_categoricas = [
    'gender', 'Partner', 'Dependents', 'Churn', 'PhoneService',
    'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup',
    'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
    'Contract', 'PaperlessBilling', 'PaymentMethod'
]

for col in columnas_categoricas:
    Datos_Norm[col] = Datos_Norm[col].astype('category')

In [ ]:
# Analizando el uso de memoria con el cambio de tipo de dato
Datos_Norm.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   customerID        7267 non-null   string  
 1   Churn             7267 non-null   category
 2   gender            7267 non-null   category
 3   SeniorCitizen     7267 non-null   bool    
 4   Partner           7267 non-null   category
 5   Dependents        7267 non-null   category
 6   tenure            7267 non-null   int64   
 7   PhoneService      7267 non-null   category
 8   MultipleLines     7267 non-null   category
 9   InternetService   7267 non-null   category
 10  OnlineSecurity    7267 non-null   category
 11  OnlineBackup      7267 non-null   category
 12  DeviceProtection  7267 non-null   category
 13  TechSupport       7267 non-null   category
 14  StreamingTV       7267 non-null   category
 15  StreamingMovies   7267 non-null   category
 16  Contract          7267 n

#### Manejo de Columnas

##### Estandarizacion de caracteres

In [232]:
Datos_Norm.columns = Datos_Norm.columns.str.lower().str.replace(' ', '_')
Datos_Norm.head()

customerid churn  gender  seniorcitizen partner dependents  tenure  \
0  0002-ORFBO    No  Female          False     Yes        Yes       9   
1  0003-MKNFE    No    Male          False      No         No       9   
2  0004-TLHLJ   Yes    Male          False      No         No       4   
3  0011-IGKFF   Yes    Male           True     Yes         No      13   
4  0013-EXCHZ   Yes  Female           True     Yes         No       3   

  phoneservice multiplelines internetservice  ... onlinebackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  deviceprotection techsupport streamingtv streamingmovies        contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  paperlessbilling     paymentmethod charges.monthly  charges.total  
0              Yes      Mailed check            65.6         593.30  
1               No      Mailed check            59.9         542.40  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9         267.40  

[5 rows x 21 columns]

##### Verificacion de datos nulos

In [233]:
valores_nulos = Datos_Norm.isna().values.sum() 
print(f"Valores nulos por columna:\n {col} {valores_nulos [valores_nulos > 0]}") 


Valores nulos por columna:
 charges.total []


In [234]:
for col in Datos_Norm.columns:
    vacios = Datos_Norm[col].isna().values.sum()# Cuenta los valores exactamente iguales a ""
    if vacios > 0:
        print(f" • {col}: {vacios} valores nulos")
        print(Datos_Norm[Datos_Norm[col].isna()][col].sample(5))

In [235]:
Datos_Norm.iloc[2930]

customerid            4075-WKNIU
churn                         No
gender                    Female
seniorcitizen              False
partner                      Yes
dependents                   Yes
tenure                         0
phoneservice                 Yes
multiplelines                Yes
internetservice              DSL
onlinesecurity                No
onlinebackup                 Yes
deviceprotection             Yes
techsupport                  Yes
streamingtv                  Yes
streamingmovies               No
contract                Two year
paperlessbilling              No
paymentmethod       Mailed check
charges.monthly            73.35
charges.total                0.0
Name: 2930, dtype: object

Pareciera que los Nan y el Valor de 0 en la columna 'tenure' coincide.

#### Consulta de tenure y los valores Nan

Prueba

Aqui vemos que la relacion de tenure = 1
quiere decir que las columnas 'charges.monthly' y 'charges.total' son iguales

In [236]:
tenure_1 = Datos_Norm[Datos_Norm['tenure'] == 1]
tenure_1.head(3)

customerid churn  gender  seniorcitizen partner dependents  tenure  \
17  0021-IKXGC    No  Female           True      No         No       1   
19  0023-HGHWL   Yes    Male           True      No         No       1   
25  0032-PGELS   Yes  Female          False     Yes        Yes       1   

   phoneservice     multiplelines internetservice  ... onlinebackup  \
17          Yes               Yes     Fiber optic  ...           No   
19           No  No phone service             DSL  ...           No   
25           No  No phone service             DSL  ...           No   

   deviceprotection techsupport streamingtv streamingmovies        contract  \
17               No          No          No              No  Month-to-month   
19               No          No          No              No  Month-to-month   
25               No          No          No              No  Month-to-month   

   paperlessbilling              paymentmethod charges.monthly  charges.total  
17              Yes           Electronic check            72.1           72.1  
19              Yes           Electronic check            25.1           25.1  
25               No  Bank transfer (automatic)            30.5           30.5  

[3 rows x 21 columns]

In [237]:
# Esto es para verificar que los customerid de tenure_0 y Valor_Nan son los mismos
comparacion_equals_2 = tenure_1['charges.monthly'].equals(tenure_1['charges.total'])
comparacion_equals_2

True

Buscamos tenure con equivalencia a 0

In [238]:
tenure_0 = Datos_Norm[Datos_Norm['tenure'] == 0]
tenure_0.head(11)

customerid churn  gender  seniorcitizen partner dependents  tenure  \
975   1371-DWPAZ    No  Female          False     Yes        Yes       0   
1775  2520-SGTTA    No  Female          False     Yes        Yes       0   
1955  2775-SEFEE    No    Male          False      No        Yes       0   
2075  2923-ARZLG    No    Male          False     Yes        Yes       0   
2232  3115-CZMZD    No    Male          False      No        Yes       0   
2308  3213-VVOLG    No    Male          False     Yes        Yes       0   
2930  4075-WKNIU    No  Female          False     Yes        Yes       0   
3134  4367-NUYAO    No    Male          False     Yes        Yes       0   
3203  4472-LVYGI    No  Female          False     Yes        Yes       0   
4169  5709-LVOEQ    No  Female          False     Yes        Yes       0   
5599  7644-OMVMY    No    Male          False     Yes        Yes       0   

     phoneservice     multiplelines internetservice  ...         onlinebackup  \
975            No  No phone service             DSL  ...                  Yes   
1775          Yes                No              No  ...  No internet service   
1955          Yes               Yes             DSL  ...                  Yes   
2075          Yes                No              No  ...  No internet service   
2232          Yes                No              No  ...  No internet service   
2308          Yes               Yes              No  ...  No internet service   
2930          Yes               Yes             DSL  ...                  Yes   
3134          Yes               Yes              No  ...  No internet service   
3203           No  No phone service             DSL  ...                   No   
4169          Yes                No             DSL  ...                  Yes   
5599          Yes                No              No  ...  No internet service   

         deviceprotection          techsupport          streamingtv  \
975                   Yes                  Yes                  Yes   
1775  No internet service  No internet service  No internet service   
1955                   No                  Yes                   No   
2075  No internet service  No internet service  No internet service   
2232  No internet service  No internet service  No internet service   
2308  No internet service  No internet service  No internet service   
2930                  Yes                  Yes                  Yes   
3134  No internet service  No internet service  No internet service   
3203                  Yes                  Yes                  Yes   
4169                  Yes                   No                  Yes   
5599  No internet service  No internet service  No internet service   

          streamingmovies  contract paperlessbilling  \
975                    No  Two year               No   
1775  No internet service  Two year               No   
1955                   No  Two year              Yes   
2075  No internet service  One year              Yes   
2232  No internet service  Two year               No   
2308  No internet service  Two year               No   
2930                   No  Two year               No   
3134  No internet service  Two year               No   
3203                   No  Two year              Yes   
4169                  Yes  Two year               No   
5599  No internet service  Two year               No   

                  paymentmethod charges.monthly  charges.total  
975     Credit card (automatic)           56.05            0.0  
1775               Mailed check           20.00            0.0  
1955  Bank transfer (automatic)           61.90            0.0  
2075               Mailed check           19.70            0.0  
2232               Mailed check           20.25            0.0  
2308               Mailed check           25.35            0.0  
2930               Mailed check           73.35            0.0  
3134               Mailed check           25.75            0.0  
3203  Ban

In [239]:
Valor_Nan = Datos_Norm[Datos_Norm['charges.total'].isna()]
Valor_Nan.head(11)

Empty DataFrame
Columns: [customerid, churn, gender, seniorcitizen, partner, dependents, tenure, phoneservice, multiplelines, internetservice, onlinesecurity, onlinebackup, deviceprotection, techsupport, streamingtv, streamingmovies, contract, paperlessbilling, paymentmethod, charges.monthly, charges.total]
Index: []

[0 rows x 21 columns]

In [240]:
# Esto es para verificar que los customerid de tenure_0 y Valor_Nan son los mismos
comparacion_equals = tenure_0['customerid'].equals(Valor_Nan['customerid'])
comparacion_equals

False

Aun no se ha cobrado el servicio (Tenure = 0), por lo que no hay un valor de cobro total (NaN).
Por lo tanto, no es necesario eliminar estos valores, ya que son válidos en el dataset.
Así que, lo conveniente seria tratar al valor como 0 en el valor de cobro total.

#### Tratamiento de NaN

Se añadio en la parte de 'Manejo de dato', un extra que al momento de normalizar, si encuentra un valor NaN lo reemplaze por "0"

```
 Datos_Norm['Charges.Total'] = pd.to_numeric(Datos_Norm['Charges.Total'], errors='coerce').fillna(0)
```

#### Arreglo de "" en 'churn'

Segun la comprobacion de Datos en Valores Vacios ("")<br>• Churn: Tiene 224 valores vacíos ('')


##### Comprobar de datos ("") en Churn

In [197]:
Churn_a = Datos_Norm[Datos_Norm['churn'] == '']
Churn_a.sample(3)

customerid churn  gender  seniorcitizen partner dependents  tenure  \
5341  7295-EXSMO        Female           True      No         No       3   
3688  5083-QPLHM          Male          False     Yes        Yes      14   
6866  9477-UIFFV          Male          False      No         No       1   

     phoneservice multiplelines internetservice  ...         onlinebackup  \
5341          Yes            No     Fiber optic  ...                   No   
3688          Yes            No             DSL  ...                  Yes   
6866          Yes            No              No  ...  No internet service   

         deviceprotection          techsupport          streamingtv  \
5341                  Yes                   No                   No   
3688                   No                   No                  Yes   
6866  No internet service  No internet service  No internet service   

          streamingmovies        contract paperlessbilling     paymentmethod  \
5341                   No  Month-to-month               No      Mailed check   
3688                   No  Month-to-month              Yes  Electronic check   
6866  No internet service  Month-to-month               No      Mailed check   

     charges.monthly  charges.total  
5341           75.05          202.9  
3688           65.45          937.6  
6866           20.30           20.3  

[3 rows x 21 columns]

In [163]:
for col in Datos_Norm.columns:
        vacios = (Datos_Norm[col] == "").sum()  # Cuenta los valores exactamente iguales a ""
        #Si el numero de valores vacios es mayor a 0, muestra ejemplos de los datos vacios
        if vacios > 0:
            print(f" • {col}: {vacios} valores vacíos ('')")
            print(f"Las Filas tienen datos vacios son: ", Datos_Norm[Datos_Norm[col] == ""][col].head(3))  # Muestra ejemplos si hay vacíos
            print('-' * 50)

 • churn: 224 valores vacíos ('')
Las Filas tienen datos vacios son:  30    
75    
96    
Name: churn, dtype: category
Categories (3, object): ['', 'No', 'Yes']
--------------------------------------------------


##### Arreglo de ("")

¿Por qué se hizo esta función?<br>

En muchos registros del dataset de clientes, la columna `churn` (abandono) aparece vacía.<br>
Para poder analizar y modelar estos datos correctamente, es importante completar esos valores.<br>

Esta función utiliza una **estrategia heurística**, es decir, una lógica basada en el sentido común<br>
y patrones observados en los datos reales:<br>

- Clientes nuevos y sin contrato a largo plazo tienden a irse más fácilmente.
- El pago con cheque electrónico suele asociarse con menor fidelización.
- El alto costo mensual sin beneficios adicionales puede aumentar la probabilidad de cancelación.

La función evalúa esas condiciones y asigna `"Yes"` (abandonó) o `"No"` (sigue) a los clientes<br>
que no tenían etiqueta, sin necesidad de usar algoritmos complejos de aprendizaje automático.<br>

Este enfoque es útil como solución inicial o cuando no se dispone de un modelo entrenado.

In [243]:
def Asignacion_de_Churn(datos):
    datos = datos.copy()  # Se hace una copia para no modificar el DataFrame original

    # Calculamos el tercer cuartil de gastos mensuales
    # Esto nos sirve como referencia para saber qué clientes pagan mucho
    umbral_gasto_alto = datos['charges.monthly'].quantile(0.75)

    # Definimos una función interna que evalúa si un cliente esta en "riesgoso"
    # Suma condiciones y marca como "Sí" (abandona) si se cumplen 3 o más
    def evaluar_riesgo(fila):
        condiciones_riesgo = 0

        # Cliente con contrato mes a mes → fácil cancelación
        if fila['contract'] == 'Month-to-month':
            condiciones_riesgo += 1
        # Pago manual con cheque electrónico (Electronic check) → baja fidelización
        if fila['paymentmethod'] == 'Electronic check':
            condiciones_riesgo += 1
        # Antigüedad corta (≤ 6 meses) → cliente aún no afianzado
        if fila['tenure'] <= 6:
            condiciones_riesgo += 1
        # Fibra óptica sin protección de seguridad ('onlinesecurity') → alto coste sin valor añadido
        if fila['internetservice'] == 'Fiber optic' and fila['onlinesecurity'] == 'No':
            condiciones_riesgo += 1
        # Gasto mensual muy alto, por encima del promedio 'Q3'
        if fila['charges.monthly'] > umbral_gasto_alto:
            condiciones_riesgo += 1

        # No tiene ningún servicio adicional (de valor agregado)
        extras = ['onlinesecurity', 'onlinebackup', 'deviceprotection', 'techsupport']
        if all(fila[e] in ['No', 'No internet service'] for e in extras):
            condiciones_riesgo += 1

        # Hace streaming (TV o Películas) con DSL → mala experiencia
        if fila['internetservice'] == 'DSL' and (
            fila['streamingtv'] == 'Yes' or fila['streamingmovies'] == 'Yes'):
            condiciones_riesgo += 1
        # Cliente mayor SIN soporte técnico → posible frustración
        if fila['seniorcitizen'] == 1 and fila['techsupport'] == 'No':
            condiciones_riesgo += 1
    
        # Si cumple 4 o más condiciones de riesgo → se marca como que abandonó (churn = 'Yes')
        return 'Yes' if condiciones_riesgo >= 4 else 'No'

    # Aplicar la lógica solo a los registros que tienen el campo vacío
    ChurnVacio = datos['churn'] == ""
    datos.loc[ChurnVacio, 'churn'] = datos[ChurnVacio].apply(evaluar_riesgo, axis=1)

    return datos

In [214]:
print(Datos_Norm['churn'].value_counts(dropna=False))  # cómo está antes

churn
No     5174
Yes    1869
        224
Name: count, dtype: int64


DATAFRAME con Churn completo

In [244]:
Churn_completo = Asignacion_de_Churn(Datos_Norm)

Verificacion y Almacenamiento de df

In [246]:
print(Churn_completo['churn'].value_counts(dropna=False))  # cómo está antes

churn
No     5339
Yes    1928
          0
Name: count, dtype: int64


In [245]:
Churn_completo.head()

customerid churn  gender  seniorcitizen partner dependents  tenure  \
0  0002-ORFBO    No  Female          False     Yes        Yes       9   
1  0003-MKNFE    No    Male          False      No         No       9   
2  0004-TLHLJ   Yes    Male          False      No         No       4   
3  0011-IGKFF   Yes    Male           True     Yes         No      13   
4  0013-EXCHZ   Yes  Female           True     Yes         No       3   

  phoneservice multiplelines internetservice  ... onlinebackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  deviceprotection techsupport streamingtv streamingmovies        contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  paperlessbilling     paymentmethod charges.monthly  charges.total  
0              Yes      Mailed check            65.6         593.30  
1               No      Mailed check            59.9         542.40  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9         267.40  

[5 rows x 21 columns]

In [247]:
Churn_completo.dtypes

customerid          string[python]
churn                     category
gender                    category
seniorcitizen                 bool
partner                   category
dependents                category
tenure                       int64
phoneservice              category
multiplelines             category
internetservice           category
onlinesecurity            category
onlinebackup              category
deviceprotection          category
techsupport               category
streamingtv               category
streamingmovies           category
contract                  category
paperlessbilling          category
paymentmethod             category
charges.monthly            float64
charges.total              float64
dtype: object

In [248]:
pd.DataFrame.to_json(Churn_completo, path_or_buf='TelecomX_Data_Churn.json', orient='records', lines=True)

## 📊 Carga y análisis

## 📄Informe final